### PR TITLE
Finish implementing code serialization support

### PR DESCRIPTION
**Pull request by dolio**
This is a pull request to complete the code saving/reloading support in the new runtime. The API provided is the one designed in #1728 

I haven't added any tests yet, as I'm not sure what sort of coverage is desirable. I'll at least add some simple ones for trivial behavior, though.

### CODE REVIEW

**Comment by pchiusano** [comment]
Exciting!

In general I'd try to do round trip testing for a bunch of definitions. I don't think you need to go crazy as a small number of definitions will exercise all the code paths of the serialization and deserialization logic.

This could be done with a transcript if we have a way of doing I/O tests with transcripts, which @stew might have some thoughts on.

File: unison-src/transcripts/alias-many.output.md
Comment (pchiusano) [inline unison-src/transcripts/alias-many.output.md]: Noting this signature is off based on earlier comment on signature given in `Builtin.hs`.

File: parser-typechecker/src/Unison/Builtin.hs
Comment (pchiusano) [inline parser-typechecker/src/Unison/Builtin.hs]: The definition should have type `[(Link.Term,Code)] ->{IO} [Link.Term]`. Tuples are right-nested pairs terminated by a `()`.

Just use this handy helper function in `Builtin.hs`. It's used by `Text.uncons` if you want to see an example. 

```suggestion
  , ("Code.cache_", list (tuple [termLink, code]) --> io (list termLink))
```

```Haskell
tuple :: Ord v => [Type v] -> Type v
tuple [t] = t
tuple ts = foldr pair (DD.unitType ()) ts
```

File: parser-typechecker/src/Unison/Runtime/ANF/Serialize.hs
Comment (pchiusano) [inline parser-typechecker/src/Unison/Runtime/ANF/Serialize.hs:4]: This file must have been fun to write. :)

**Review by pchiusano — APPROVED** [review]
LGTM. Reviewing this I am realizing we need more high level docs for runtime implementation stuff.

Before merging, can you fix the signature on `cache_`? See my inline comment.

I'd probably work on some tests part of this PR, but if you want to merge this sooner and create a ticket for adding tests and get on that ASAP that's okay with me.

@stew am a little worried this PR will be a pain to reconcile with your branch you're working on.

File: parser-typechecker/src/Unison/Runtime/ANF/Serialize.hs
Comment (dolio) [reply parser-typechecker/src/Unison/Runtime/ANF/Serialize.hs:4]: Super fun.

File: unison-src/new-runtime-transcripts/codeops.md
Comment (pchiusano) [inline unison-src/new-runtime-transcripts/codeops.md]: Lol. Epic. I guess fine for now.

File: unison-src/new-runtime-transcripts/codeops.md
Comment (pchiusano) [inline unison-src/new-runtime-transcripts/codeops.md:9]: How about some more tests for this transcript (that round trip is `Universal.==` for a variety of definitions), following https://github.com/unisonweb/unison/blob/trunk/unison-src/new-runtime-transcripts/builtins.output.md for inspiration. 

Basically, you create a bunch of `test>` watch expressions, add them all to the codebase, then do one `.> test` at the end.

**Comment by dolio** [comment]
FYI: don't merge this yet. The transcript I wrote used an IO loophole that won't work once the 1696 fix is merged.

File: unison-src/new-runtime-transcripts/codeops.md
Comment (dolio) [reply unison-src/new-runtime-transcripts/codeops.md:9]: I've added some more complexity to the thing that's serialized. I can't run deserialization in a test block, though, because it involves IO, no?

Comment (pchiusano) [reply unison-src/new-runtime-transcripts/codeops.md:9]: That's right. 

@stew has an implementation of a `io.test` command that lets you run I/O-based tests, so let's leave this until that merges. I'll create an issue to track.

**Pull request merged**
